addons/better-terrain/editor/TileView.gd

```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
+	if clicked:
+		initial_click = current_position
+		selection_start = Vector2i(-1,-1)
+	if released:
+		selection_rect = Rect2i(0,0,0,0)
+		queue_redraw()
 	
-	if (clicked or event is InputEventMouseMotion) and paint_action != PaintAction.NO_ACTION:
-		if !highlighted_tile_part.valid:
-			return
+	if paint_action == PaintAction.PASTE:
+		if event is InputEventMouseMotion:
+			queue_redraw()
 		
-		if paint_action == PaintAction.DRAW_TYPE or paint_action == PaintAction.ERASE_TYPE:
-			var type := BetterTerrain.get_tile_terrain_type(highlighted_tile_part.data)
-			var goal := paint if paint_action == PaintAction.DRAW_TYPE else -1
-			if type != goal:
-				undo_manager.create_action("Set tile terrain type", UndoRedo.MERGE_DISABLE, tileset)
-				undo_manager.add_do_method(BetterTerrain, &"set_tile_terrain_type", tileset, highlighted_tile_part.data, goal)
-				undo_manager.add_do_method(self, &"queue_redraw")
-				if goal == -1:
-					terrain_undo.create_peering_restore_point_tile(
-						undo_manager,
-						tileset,
-						highlighted_tile_part.source_id,
-						highlighted_tile_part.coord,
-						highlighted_tile_part.alternate
-					)
-				else:
-					undo_manager.add_undo_method(BetterTerrain, &"set_tile_terrain_type", tileset, highlighted_tile_part.data, type)
-				undo_manager.add_undo_method(self, &"queue_redraw")
-				undo_manager.commit_action()
-		elif paint_action == PaintAction.DRAW_PEERING:
-			if highlighted_tile_part.has("peering"):
-				if !(paint in BetterTerrain.tile_peering_types(highlighted_tile_part.data, highlighted_tile_part.peering)):
-					undo_manager.create_action("Set tile terrain peering type", UndoRedo.MERGE_DISABLE, tileset)
-					undo_manager.add_do_method(BetterTerrain, &"add_tile_peering_type", tileset, highlighted_tile_part.data, highlighted_tile_part.peering, paint)
+		if clicked:
+			if event.button_index == MOUSE_BUTTON_LEFT:
+				if staged_paste_tile_states.size() > 0:
+					undo_manager.create_action("Paste tile terrain peering types", UndoRedo.MERGE_DISABLE, tileset)
+					var base_rect = staged_paste_tile_states[0].base_rect
+					for p in staged_paste_tile_states:
+						var staged_rect:Rect2 = p.base_rect
+						staged_rect.position -= base_rect.position + base_rect.size / 2
+						
+						staged_rect.position *= zoom_level
+						staged_rect.size *= zoom_level
+						
+						staged_rect.position += Vector2(current_position)
+						
+						var old_tile_part = tile_part_from_position(staged_rect.get_center())
+						var new_tile_state = p
+						if old_tile_part.valid and new_tile_state.part.valid:
+							for side in range(16): #not sure how to loop over or get max value of an enum like TileSet.CellNeighbor
+								var old_peering = BetterTerrain.tile_peering_types(old_tile_part.data, side)
+								var new_sides = new_tile_state.sides
+								if old_peering.has(paint) != new_sides.has(side):
+									if new_sides.has(side):
```

Comment:
I'd use `if new_sides.has(side) and not old_peering.has(paint):` and for the else, `elif old_peering.has(paint) and not new_sides.has(side):`. That would be easier to read for me, and less indented.